src/jit/decomposelongs.cpp

```diff
@@ -987,77 +977,55 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
 {
     assert(use.IsInitialized());
 
-    GenTree* tree         = use.Def();
-    GenTree* gtLong       = tree->gtGetOp1();
-    GenTree* oldShiftByOp = tree->gtGetOp2();
+    GenTree* tree      = use.Def();
+    GenTree* gtLong    = tree->gtGetOp1();
+    GenTree* loOp1     = gtLong->gtGetOp1();
+    GenTree* hiOp1     = gtLong->gtGetOp2();
+    GenTree* shiftByOp = tree->gtGetOp2();
 
     genTreeOps oper        = tree->OperGet();
-    genTreeOps shiftByOper = oldShiftByOp->OperGet();
+    genTreeOps shiftByOper = shiftByOp->OperGet();
 
     assert((oper == GT_LSH) || (oper == GT_RSH) || (oper == GT_RSZ));
 
-    unsigned loOp1LclNum;
-    unsigned hiOp1LclNum;
-
-    if (gtLong->gtOp.gtOp1->OperGet() != GT_LCL_VAR)
-    {
-        LIR::Use loOp1Use(Range(), &gtLong->gtOp.gtOp1, gtLong);
-        loOp1LclNum = loOp1Use.ReplaceWithLclVar(m_compiler, m_blockWeight);
-    }
-    else
-    {
-        loOp1LclNum = gtLong->gtOp.gtOp1->AsLclVarCommon()->gtLclNum;
-    }
-
-    if (gtLong->gtOp.gtOp2->OperGet() != GT_LCL_VAR)
-    {
-        LIR::Use hiOp1Use(Range(), &gtLong->gtOp.gtOp2, gtLong);
-        hiOp1LclNum = hiOp1Use.ReplaceWithLclVar(m_compiler, m_blockWeight);
-    }
-    else
-    {
-        hiOp1LclNum = gtLong->gtOp.gtOp2->AsLclVarCommon()->gtLclNum;
-    }
-
-    GenTree* loOp1 = gtLong->gtGetOp1();
-    GenTree* hiOp1 = gtLong->gtGetOp2();
-
-    Range().Remove(gtLong);
-    Range().Remove(loOp1);
-    Range().Remove(hiOp1);
-
     // If we are shifting by a constant int, we do not want to use a helper, instead, we decompose.
     if (shiftByOper == GT_CNS_INT)
     {
-        unsigned int count = oldShiftByOp->gtIntCon.gtIconVal;
-        Range().Remove(oldShiftByOp);
+        unsigned int count = shiftByOp->gtIntCon.gtIconVal;
+        Range().Remove(shiftByOp);
+
+        if (count == 0)
+        {
+            // Remove tree and don't do anything else.
+            Range().Remove(tree);
+
+            use.ReplaceWith(m_compiler, gtLong);
+            return gtLong->gtNext;
```

Comment:
This may end up performing redundant processing on nodes that have already been decomposed. You should save `tree->gtNext` before removing `tree` and return the saved value.